### PR TITLE
Add support for QP solving with Clarabel

### PR DESCRIPTION
**Pull request by isaacOnline**
### Description:
This PR adds support for the Clarabel quadratic program solver. When using similar settings, Clarabel is consistently faster than OSQP across a wide range of test programs, often by 50% or more. It also tends to find more accurate solutions for difficult problems (see [qpbenchmark](https://github.com/qpsolvers/qpbenchmark)).

### Changes:
- Introduces `solver` parameters to `pensynth` and `cv_pensynth`, allowing users to choose between osqp and clarabel. 
- Introduces `return_solver_info` parameter to `cv_pensynth`, which results in diagnostic information about the optimization process being returned. 
- Adds documentation for new parameters. 
- Does not change default behavior of `pensynth` and `cv_pensynth`. 

Happy to modify the PR if need be, so let me know if there's anything I should change!

### CODE REVIEW

**Comment by vankesteren** [comment]
Hey @isaacOnline, thanks so much for your PR! Very nice that you want to contribute. Here are my overall comments, mostly based on my limited time available for software maintenance. Any contribution should be very stable

- I want to keep the dependencies to a minimum. If we do choose to allow for different solvers, we should do so in the "suggests" field of the description and check if the package is installed at runtime when the alternative solver is chosen.
- The infrastructure for allowing different solvers adds quite some complexity to what is supposed to be a simple package. There needs to be a good argument for allowing this, and if we do choose to do this I think we should slightly update the architecture: create a separate file `solvers.R` with different functions that adhere to the same interface `solve_*(X0VX0, X1VX0, Delta, lambda)`:
  - `solve_osqp()`
  - `solve_clarabel()`
  - Then it will be easy to add `solve_fancynewQPsolver()` in the future
- What problem does this solve exactly for pensynth? And if Clarabel is so much better than OSQP, why not just replace OSQP and keep the package simple?
- To partially answer my previous point, Clarabel itself seems to be quite a new project. In general, osqp seems more mature. I really like it, but also it's an order of magnitude less popular, which generally translates to less support:
  - [](https://CRAN.R-project.org/package=clarabel) for clarabel 
  - [](https://CRAN.R-project.org/package=osqp) for osqp

In conclusion, I am definitely open in principle to either integrating additional solvers or replacing osqp, but only if:
- it solves an actual problem (and maybe you indeed had a problem with running a certain penalized synthetic control!!), rather than potential benchmark improvements
- the additional benefit outweighs the additional complexity.

I also say this because next on my to-do list for this package is enabling multiple treated units, which should increase the complexity a bit already.

**Comment by vankesteren** [comment]
PS: just to add, I really like some of the stuff you did here such as allowing solver info to be returned.

File: R/cv_pensynth.R [hunk 1]
```diff
@@ -57,38 +65,125 @@ cv_pensynth <- function(X1, X0, v, Z1, Z0, nlambda = 100, opt_pars = osqp::osqpS
...
+    # Set default parameters for clarabel
+    if (is.null(opt_pars)) {
+      opt_pars <- clarabel::clarabel_control()
+    }
+
+    # Constraint matrices
+    Amat <- rbind(
+      rep(1, N_donors), # Sum to 1 constraint
+      diag(N_donors), # Individ. weights lte 1 constraint
+      -diag(N_donors) # Individ. weights gte 0 constraint
```
Comment (vankesteren) [inline R/cv_pensynth.R]: We would only need sum to 1 and greater than 0 right? Because that automatically implies individual weights are less than 1.

**Comment by isaacOnline** [comment]
Hi @vankesteren—Thanks for the feedback! I introduced Clarabel to solve two problems that came up for me with OSQP: 1) when using default settings, it returned negative weights, despite reporting back a "solved" status, and 2) when using higher accuracy settings, it was not fast enough to work with my dataset. 

You can recreate the negative weights issue using the example in your readme, where it looks like about a quarter of the weights come out as negative (`mean(res$w_path < 0)`), and together the negative weights when using the cv-selected lambda sum to about -0.001 (`sum(res$w_opt[res$w_opt < 0])`).  I’ve actually had lower weights sums (up to about -0.05) in the dataset I’m using. 

I tried to fix this by using stricter optimization settings on OSQP, but found the solver to be slower when doing so. I included a snippet below for timing a similar sized problem to the one I’m working on for a single treated unit, where it took my laptop ~10.4 min to solve with OSQP and ~5.1 min to solve with Clarabel. 

If I were to pick one of the solvers, the speed savings are enough for me that switching to Clarabel makes sense, but I definitely understand it isn’t worth the risk of less maintainability to you. (My dataset has ~500 treated units and between 500 and 3,500 potential controls for each treated case, so I’m fairly sensitive to speed.) If you don't want to switch, I'd suggest at least checking and raising warnings for when weights are negative as otherwise this behavior won’t be surfaced. You might also consider changing the default OSQP settings to make them a bit higher accuracy. 

Let me know how you’d like to proceed and I’m happy to make changes.

```
remotes::install_github("isaacOnline/pensynth")
library(pensynth)
set.seed(45)

# Generate some data
N_covar  <- 1   # number of covariates
N_donor  <- 1000  # number of donor units
N_target <- 40 # pre-intervention time series length
w  <- runif(N_donor) # true unit weights
w[5:N_donor] <- 0    # set most to 0 (sparse)
w  <- w / sum(w)     # normalize unit weights
v  <- rep(1, N_covar) # equal variable weights

# Create the synthetic control data matrices
# following the Synth::synth() notation
X0 <- matrix(rnorm(N_covar*N_donor), N_covar)  
X1 <- X0 %*% w
Z0 <- matrix(rnorm(N_target*N_donor), N_target)
Z1 <- Z0 %*% w


# Measure OSQP Speed
start.time <- Sys.time()
res <- cv_pensynth(X1, X0, v, Z1, Z0, solver="osqp", opt_pars=osqp::osqpSettings(eps_abs=1e-09, eps_rel=0.0))
end.time <- Sys.time()
time.taken <- end.time - start.time
print("OSQP: ")
print(time.taken)

# Measure Clarabel Speed
start.time <- Sys.time()
res <- cv_pensynth(X1, X0, v, Z1, Z0, solver="clarabel")
end.time <- Sys.time()
time.taken <- end.time - start.time
print("Clarabel: ")
print(time.taken)
```

**Comment by vankesteren** [comment]
Alright, I tested it too and that sounds like a pretty good reason to me to switch to clarabel completely!

I don't think we need the infrastructure for different solvers for now. Let's remove osqp then and bump the version to 0.3.0. (and we should also update the Readme accordingly!) You can do any sub-parts of this, I can help finish it as well.

**Comment by isaacOnline** [comment]
Just made the switch. I updated the documentation and readme as well. It looks like the test for `pensynth` is failing now, but I wasn't able to figure out what was going on there.

File: R/pensynth.R [hunk 2]
```diff
@@ -73,30 +72,34 @@ pensynth <- function(X1, X0, v, lambda = 0, opt_pars = osqp::osqpSettings(polish
...
+  )
+  B <- c(
+    1, # Sum to 1 constraint
+    rep(0, N_donors) # Individ. weights gte 0 constraint
+  )
 
-  # stop annoying printing with capture.output.
-  o <- capture.output({
+# Run the quadratic program solver
+  result <- clarabel(
```
Comment (vankesteren) [inline R/pensynth.R]:
```suggestion
  result <- clarabel::clarabel(
```
This will fix the error!

File: DESCRIPTION [hunk 3]
```diff
@@ -1,7 +1,7 @@
 Package: pensynth
 Type: Package
 Title: Penalized Synthetic Control Estimation
-Version: 0.2.0
+Version: 0.3.0
 Author: Erik-Jan van Kesteren <e.vankesteren1@uu.nl>
```
Comment (vankesteren) [inline DESCRIPTION]: You can add yourself as contributor!

**Review by vankesteren — CHANGES_REQUESTED** [review]
Comment for the CV MSE plot in the readme: it would be nice if it were not monotonically increasing. For illustrative purposes, let's put an example that actually requires some nonzero lambda penalty. Did you run the example with a seed of 45? Because it used to be like that!

EDIT: never mind this, i just checked a bit better and the new version goes to the same  solution as the old version with a larger penalty range. I'll look at this after we merge this one!

**Comment by vankesteren** [comment]
Awesome work, almost there!

File: DESCRIPTION (hunk 3 above)
Comment (isaacOnline) [reply DESCRIPTION]: Make sure the syntax is right!

**Comment by vankesteren** [comment]
Judging by the failing tests, I think the description file is still malformed with the parentheses somehow
